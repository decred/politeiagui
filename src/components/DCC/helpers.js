import {
  DCC_STATUS_ACTIVE,
  // DCC_STATUS_SUPPORTED,
  DCC_STATUS_APPROVED,
  DCC_STATUS_REJECTED,
  DCC_TYPE_ISSUANCE,
  DCC_TYPE_REVOCATION
} from "../../constants";

export const dccStatusList = [
  {
    label: "active",
    value: DCC_STATUS_ACTIVE
  },
  {
    label: "approved",
    value: DCC_STATUS_APPROVED
  },
  {
    label: "rejected",
    value: DCC_STATUS_REJECTED
  }
];

export const dccChangeStatusList = [
  // {
  //   label: "support",
  //   value: DCC_STATUS_SUPPORTED
  // },
  {
    label: "approve",
    value: DCC_STATUS_APPROVED
  },
  {
    label: "reject",
    value: DCC_STATUS_REJECTED
  }
  // {
  //   label: "debate",
  //   value: DCC_STATUS_DEBATE
  // }
];

export const typesForDCC = {
  [DCC_TYPE_ISSUANCE]: "Issuance",
  [DCC_TYPE_REVOCATION]: "Revocation"
};

export const getDCCStatus = (status) => {
  const { label } = dccStatusList.find((st) => st.value === status);
  return label;
};

const separateName = (length) => (name, index) =>
  index < length - 1 ? `${name}, ` : name;

export const getVotesUsernameList = (names) =>
  names.map(separateName(names.length));
