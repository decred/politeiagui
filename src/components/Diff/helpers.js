import isEqual from "lodash/fp/isEqual";

const markLineAsAdded = (elem) => ({
  value: elem.value,
  lineIndex: elem.index,
  removed: false,
  added: true
});

const markLineAsRemoved = (elem) => ({
  lineIndex: elem.index,
  removed: elem.value,
  added: false
});

const markLineAsUnchanged = (elem) => ({
  value: elem.value,
  lineIndex: elem.index,
  removed: false,
  added: false
});

export const arrayDiff = (newCommentBody, oldCommentBody, lineDiffFunc) => [
  ...newCommentBody.filter(lineDiffFunc(oldCommentBody)).map(markLineAsAdded),
  ...oldCommentBody.filter(lineDiffFunc(newCommentBody)).map(markLineAsRemoved),
  ...newCommentBody.filter(lineEqFunc(oldCommentBody)).map(markLineAsUnchanged)
];

export const lineDiffFunc = (arr) => (elem) =>
  !arr.some((arrelem) => arrelem.value === elem.value);

export const lineEqFunc = (arr) => (elem) => !lineDiffFunc(arr)(elem);

export const getLineArray = (string) =>
  string && string.length
    ? string.split("\n\n").map((line, index) => ({ value: line, index: index }))
    : [];

const markAsAdded = (elem) => ({ ...elem, added: true });
const markAsRemoved = (elem) => ({ ...elem, removed: true });

const filesDiffFunc = (arr) => (elem) =>
  !arr.some(
    (arrelem) => arrelem.name === elem.name && arrelem.payload === elem.payload
  );
const filesEqFunc = (arr) => (elem) => !filesDiffFunc(arr)(elem);

export const getFilesDiff = (newFiles, oldFiles) => [
  ...newFiles.filter(filesDiffFunc(oldFiles)).map(markAsAdded),
  ...oldFiles.filter(filesDiffFunc(newFiles)).map(markAsRemoved),
  ...newFiles.filter(filesEqFunc(oldFiles)) // for unchanged files
];

const itemsDiffFunc = (arr) => (elem) => !arr.some(isEqual(elem));
const itemsEqFunc = (arr) => (elem) => !itemsDiffFunc(arr)(elem);

export const getLineitemsDiff = (newItems = [], oldItems = []) =>
  [
    ...newItems.filter(itemsDiffFunc(oldItems)).map(markAsAdded),
    ...oldItems.filter(itemsDiffFunc(newItems)).map(markAsRemoved),
    ...newItems.filter(itemsEqFunc(oldItems))
  ].sort((a, b) => a.index - b.index);

export const setLineitemParams = (lineItems = [], params = {}) =>
  lineItems.map((line, index) => ({ ...params, ...line, index }));
