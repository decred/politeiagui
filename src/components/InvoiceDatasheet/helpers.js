import {
  fromHoursToMinutes,
  fromUSDCentsToUSDUnits,
  fromMinutesToHours,
  fromUSDUnitsToUSDCents
} from "../../helpers";

export const convertLineItemsToGrid = (lineItems, readOnly = true) => {
  const grid = [createTableHeaders()];
  const gridBody = lineItems.map((line, idx) => {
    const isLabelReadonly =
      line.type === 2 ? true : line.type === 3 ? true : readOnly;
    const isExpenseReadonly = line.type === 1 ? true : readOnly;
    return [
      { readOnly: true, value: idx + 1 },
      { readOnly, value: line.type },
      { readOnly, value: line.domain },
      { readOnly, value: line.subdomain },
      { readOnly, value: line.description },
      { readOnly, value: line.proposaltoken },
      { readOnly: isLabelReadonly, value: fromMinutesToHours(line.labor) },
      {
        readOnly: isExpenseReadonly,
        value: fromUSDCentsToUSDUnits(line.expenses)
      }
    ];
  });
  return grid.concat(gridBody);
};

export const convertGridToLineItems = grid => {
  const copyGrid = grid.map(row => [...row]);
  return copyGrid.reduce((acc, rowValues, row) => {
    // skip first row as it is exclusive for table headers
    if (row === 0) return acc;

    const lineItem = rowValues.reduce((acc, cell, col) => {
      switch (col) {
        case columnTypes.TYPE_COL:
          return { ...acc, type: +cell.value };
        case columnTypes.DOMAIN_COL:
          return { ...acc, domain: cell.value };
        case columnTypes.SUBDOMAIN_COL:
          return { ...acc, subdomain: cell.value };
        case columnTypes.DESC_COL:
          return { ...acc, description: cell.value };
        case columnTypes.PROP_TOKEN_COL:
          return { ...acc, proposaltoken: cell.value };
        case columnTypes.LABOR_COL:
          return { ...acc, labor: fromHoursToMinutes(+cell.value) };
        case columnTypes.EXP_COL:
          return { ...acc, expenses: fromUSDUnitsToUSDCents(+cell.value) };
        default:
          return acc;
      }
    }, {});

    return acc.concat([lineItem]);
  }, []);
};

export const columnTypes = {
  TYPE_COL: 1,
  DOMAIN_COL: 2,
  SUBDOMAIN_COL: 3,
  DESC_COL: 4,
  PROP_TOKEN_COL: 5,
  LABOR_COL: 6,
  EXP_COL: 7
};

const { TYPE_COL, LABOR_COL, EXP_COL } = columnTypes;

export const createTableHeaders = () => [
  { readOnly: true, value: "", width: 10 },
  { value: "Type", width: 40, readOnly: true },
  { value: "Domain", readOnly: true },
  { value: "Subdomain", readOnly: true },
  { value: "Description", readOnly: true },
  { value: "Proposal Token", readOnly: true },
  { value: "Labor (hours)", width: 60, readOnly: true },
  { value: "Expense (USD)", width: 60, readOnly: true }
];

export const createNewRow = rowNum => {
  return [
    { readOnly: true, value: rowNum },
    { value: "" },
    { value: "" },
    { value: "" },
    { value: "" },
    { value: "" },
    { value: "" },
    { value: "" }
  ];
};

export const errors = {
  InvalidType: 1,
  InvalidDomainLen: 2,
  InvalidSubdomainLen: 3,
  InvalidDescriptionLen: 4,
  InvalidPropTokenLen: 5,
  InvalidLaborValue: 6,
  InvalidLaborAmount: 7,
  InvalidExpenseValue: 8,
  InvalidExpenseAmount: 9
};

export const errorsMessage = {
  [errors.InvalidType]: "Type must be either 1, 2 or 3",
  [errors.InvalidDomainLen]: "Domain must have between 3 and 50 chars",
  [errors.InvalidSubdomainLen]: "Subomain must have between 3 and 50 chars",
  [errors.InvalidDescriptionLen]:
    "Description must have between 3 and 50 chars",
  [errors.InvalidPropTokenLen]: "Proposals tokens must be 64 characters long",
  [errors.InvalidLaborValue]: "Labor value must be a number",
  [errors.InvalidLaborAmount]:
    "If Type is 1, labor value must be greater than 0",
  [errors.InvalidExpenseValue]: "Expense value must be a number",
  [errors.InvalidExpenseAmount]:
    "If Type is 2, expense value must be greater than 0"
};

const response = (error, newValue) => ({
  error,
  newValue
});

export const updateGridCell = (grid, row, col, values) => {
  grid[row][col] = { ...grid[row][col], ...values };
  return grid;
};

export const processTypeColChange = (grid, { row, col, value }) => {
  const validTypeValues = [1, 2, 3];
  const valueIsValid = validTypeValues.find(v => v === +value);
  // validate if the value
  if (!valueIsValid) {
    return response(
      errors.InvalidType,
      updateGridCell(grid, row, col, {
        value: ""
      })
    );
  }

  // if value is 1, the expense column must be 0, so we set it to readOnly
  if (+value === 1) {
    grid = updateGridCell(grid, row, EXP_COL, { value: 0, readOnly: true });
  } else {
    // if it is not 1, we make sure it is not readOnly
    grid = updateGridCell(grid, row, EXP_COL, { readOnly: false });
  }

  // if value is 2, the labor column must be 0, so we set it to readOnly
  if (+value === 2 || +value === 3) {
    grid = updateGridCell(grid, row, LABOR_COL, { value: 0, readOnly: true });
  } else {
    // if it is not 2, we make sure it is not readOnly
    grid = updateGridCell(grid, row, LABOR_COL, { readOnly: false });
  }

  return response(null, updateGridCell(grid, row, col, { value }));
};

const validateStringLength = (policy, value) => {
  if (
    value.length < policy.minlineitemcollength ||
    value.length > policy.maxlineitemcollength
  ) {
    return false;
  }
  return true;
};

export const processStringColChange = (policy, grid, { row, col, value }) => {
  if (!validateStringLength(policy, value)) {
    return response(
      errors.InvalidDomainLen,
      updateGridCell(grid, row, col, {
        value: ""
      })
    );
  }

  return response(null, updateGridCell(grid, row, col, { value }));
};

export const processDomainColChange = (policy, grid, { row, col, value }) => {
  if (!validateStringLength(policy, value)) {
    return response(
      errors.InvalidDomainLen,
      updateGridCell(grid, row, col, {
        value: value
      })
    );
  }

  return response(null, updateGridCell(grid, row, col, { value }));
};

export const processSubdomainColChange = (
  policy,
  grid,
  { row, col, value }
) => {
  if (!validateStringLength(policy, value)) {
    return response(
      errors.InvalidSubdomainLen,
      updateGridCell(grid, row, col, {
        value: value
      })
    );
  }

  return response(null, updateGridCell(grid, row, col, { value }));
};

export const processDescriptionColChange = (
  policy,
  grid,
  { row, col, value }
) => {
  if (!validateStringLength(policy, value)) {
    return response(
      errors.InvalidDescriptionLen,
      updateGridCell(grid, row, col, {
        value: value
      })
    );
  }

  return response(null, updateGridCell(grid, row, col, { value }));
};

export const processPropTokenColChange = (grid, { row, col, value }) => {
  const updatedGrid = updateGridCell(grid, row, col, {
    value: value
  });
  // proposal token can be either blank or have a valid token
  if (!value) {
    return response(null, updatedGrid);
  }

  // for now only validate the token length
  if (value.length < 64) {
    return response(errors.InvalidPropTokenLen, updatedGrid);
  }

  return updatedGrid;
};

export const processLaborColChange = (grid, { row, col, value }) => {
  const floatValue = parseFloat(value, 10).toFixed(2);

  // make sure the value can be converted to a valid number
  const invalidNumber = !floatValue || isNaN(floatValue);
  if (invalidNumber) {
    return response(
      errors.InvalidLaborValue,
      updateGridCell(grid, row, col, {
        value: ""
      })
    );
  }

  // make sure labor is different than 0 if type is equal 1
  const typeColValue = grid[row][TYPE_COL].value;
  const isTypeOne = parseInt(typeColValue, 10) === 1;
  if (isTypeOne && +floatValue <= 0) {
    return response(
      errors.InvalidLaborAmount,
      updateGridCell(grid, row, col, {
        value: 1
      })
    );
  }

  return response(
    null,
    updateGridCell(grid, row, col, {
      value: floatValue
    })
  );
};

export const processExpenseColChange = (grid, { row, col, value }) => {
  const floatValue = parseFloat(value, 10).toFixed(2);

  const invalidNumber = !floatValue || isNaN(floatValue);

  if (invalidNumber) {
    return response(
      errors.InvalidExpenseValue,
      updateGridCell(grid, row, col, {
        value: ""
      })
    );
  }

  // make sure expense is different than 0 if type is equal 2
  const typeColValue = grid[row][TYPE_COL].value;
  const isTypeTwo = parseInt(typeColValue, 10) === 2;
  if (isTypeTwo && +floatValue <= 0) {
    return response(
      errors.InvalidExpenseAmount,
      updateGridCell(grid, row, col, {
        value: 1
      })
    );
  }

  return response(
    null,
    updateGridCell(grid, row, col, {
      value: floatValue
    })
  );
};

export const processCellsChange = (policy, currentGrid, changes) => {
  let result = null;
  const getGridAndErrorsFromResult = (acc, result) => ({
    grid: result.newValue,
    errors: result.error ? acc.errors.add(result.error) : acc.errors
  });

  return changes.reduce(
    (acc, change) => {
      switch (change.col) {
        case columnTypes.TYPE_COL:
          result = processTypeColChange(acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.DOMAIN_COL:
          result = processDomainColChange(policy, acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.SUBDOMAIN_COL:
          result = processSubdomainColChange(policy, acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.DESC_COL:
          result = processDescriptionColChange(policy, acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.PROP_TOKEN_COL:
          result = processPropTokenColChange(acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.LABOR_COL:
          result = processLaborColChange(acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        case columnTypes.EXP_COL:
          result = processExpenseColChange(acc.grid, change);
          return getGridAndErrorsFromResult(acc, result);
        default:
          acc.grid = updateGridCell(acc.grid, change.row, change.col, {
            value: change.value
          });
          return acc;
      }
    },
    { grid: currentGrid, errors: new Set() }
  );
};
