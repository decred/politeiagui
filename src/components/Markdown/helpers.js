import React from "react";
import htmlParser from "react-markdown/plugins/html-parser";
import ModalExternalLink from "../ModalExternalLink";
import useModalContext from "src/hooks/utils/useModalContext";
import xssFilters from "xss-filters";

export const htmlParserRules = htmlParser({
  isValidNode: (node) => {
    return node.type !== "script";
  }
});

export const traverseChildren = (el, cb) => {
  const filterChildren = (c) =>
    React.Children.map(c, (child) => traverseChildren(child, cb));
  let newElement = null;
  if (el.children) {
    newElement = {
      ...el,
      children: filterChildren(el.children)
    };
  } else if (el.props && el.props.children) {
    const filteredChildren = filterChildren(el.props.children);
    newElement = {
      ...el,
      props: {
        ...el.props,
        children: filteredChildren
      }
    };
  }
  return newElement ? cb(newElement) : cb(el);
};

const isExternalLink = (link) => {
  // e.preventDefault();
  // Does this to prevent xss attacks
  const tmpLink = document.createElement("a");
  tmpLink.href = link;
  const externalLink =
    tmpLink.hostname && tmpLink.hostname !== window.top.location.hostname;

  return externalLink;
};

const LinkRenderer = ({ url, children }) => {
  const [handleOpenModal, handleCloseModal] = useModalContext();

  function onLinkClick(e) {
    if (isExternalLink(url)) {
      e.preventDefault();
      handleOpenModal(ModalExternalLink, {
        onClose: handleCloseModal,
        link: url
      });
    }
  }
  return (
    <a href={url} onClick={onLinkClick}>
      {children}
    </a>
  );
};

// Use external link renderer when images are not allowed
const imageHandler = (renderImages) => ({ src, alt }) =>
  renderImages ? (
    <img src={src} alt={alt} />
  ) : (
    <LinkRenderer url={src}>{alt}</LinkRenderer>
  );

const linkHandler = ({ href, children }) => {
  return <LinkRenderer url={href}>{children}</LinkRenderer>;
};

const handleFilterXss = (el) => {
  if (typeof el === "string") return el;
  const props = el.props;
  if (!props) {
    return el;
  }
  const newProps = {
    ...props
  };
  if (newProps.src) {
    newProps.src = xssFilters.uriInDoubleQuotedAttr(props.src);
  }
  return {
    ...el,
    props: newProps
  };
};

const rootHandler = (filterUrl) => (el) => {
  if (filterUrl) {
    el = traverseChildren(el, handleFilterXss);
  }
  const { children, ...props } = el;
  return <div {...props}>{children}</div>;
};

export const customRenderers = (renderImages, filterUrl) => {
  return {
    image: imageHandler(renderImages),
    imageReference: imageHandler(renderImages),
    link: linkHandler,
    linkReference: linkHandler,
    root: rootHandler(filterUrl)
  };
};
