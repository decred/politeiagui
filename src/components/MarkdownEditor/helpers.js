const CHAR_CODES = {
  newline: 10
};

const finderResult = (mistake, message = "") => ({mistake, message});

function iterateOverStringCharsCode(str, callback) {
  for (let i = 0, len = str.length; i < len; i++) {
    callback(str.charCodeAt(i), i);
  }
}

function tryingParagraphWithOneNewline(value) {
  const { newline } = CHAR_CODES;
  let thereAreASingleNewline = false;
  iterateOverStringCharsCode(value, (charCode, index) => {
    const nextCharCode = value.charCodeAt(index + 1);
    const prevCharCode = value.charCodeAt(index - 1);
    if (charCode === newline && nextCharCode && nextCharCode !== newline && prevCharCode !== newline ) {
      thereAreASingleNewline = true;
    }
  });
  return finderResult(thereAreASingleNewline, "You should press Enter twice to create separate paragraphs of text");
}

const mistakeFinders = [
  tryingParagraphWithOneNewline
];

export function applyMistakeFinders(value) {
  const findersResults = mistakeFinders.map(f => f(value));
  const anyMistake = findersResults.filter(r => r.mistake).length > 0;
  return ({
    findersResults,
    anyMistake
  });
}

export function getPreviewContent(className) {
  const previewContent = document.querySelector(`.${className}`).firstChild;
  const nodeList = previewContent.childNodes;
  const paragraphNodes = generateParagraphNodesArray(nodeList);
  const concatedString = generateConcatedStringFromTextNodes(paragraphNodes);
  return concatedString;
}

function generateParagraphNodesArray(nodeList) {
  const paragraphNodes = [];
  for (let node of nodeList) {
    if (node.nodeName === "P") {
      paragraphNodes.push(node);
    }
  }
  return paragraphNodes;
}

function generateConcatedStringFromTextNodes(paragraphNodes) {
  const validTextChunks = paragraphNodes.reduce((acc, node) => {
    const childrens = node.childNodes;
    for (let child of childrens) {
      if (child.nodeName === "#text") {
        acc += child.nodeValue;
      }
    }
    return acc;
  }, "");
  return validTextChunks;
}
