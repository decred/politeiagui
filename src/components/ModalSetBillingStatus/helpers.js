import {
  PROPOSAL_BILLING_STATUS_ACTIVE,
  PROPOSAL_BILLING_STATUS_COMPLETED,
  PROPOSAL_BILLING_STATUS_CLOSED
} from "src/constants";

const billingStatusLabels = {
  [PROPOSAL_BILLING_STATUS_ACTIVE]: "Active",
  [PROPOSAL_BILLING_STATUS_COMPLETED]: "Completed",
  [PROPOSAL_BILLING_STATUS_CLOSED]: "Closed"
};

/**
 * Returns the proposal billing status select options.
 * @returns {Array}
 */
export const getProposalBillingStatusOptionsForSelect = () =>
  [
    PROPOSAL_BILLING_STATUS_ACTIVE,
    PROPOSAL_BILLING_STATUS_COMPLETED,
    PROPOSAL_BILLING_STATUS_CLOSED
  ].map((value) => ({
    label: billingStatusLabels[value],
    value
  }));
