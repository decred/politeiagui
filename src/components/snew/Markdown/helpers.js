import React from "react";
import xssFilters from "xss-filters";
import * as modalTypes from "../../Modal/modalTypes";

export const traverseChildren = (el, cb) => {
  const filterChildren = (c) => React.Children.map(c,
    child => traverseChildren(child, cb)
  );
  let newElement = null;
  if(el.children) {
    newElement = {
      ...el,
      children: filterChildren(el.children)
    };
  } else if (el.props && el.props.children) {
    const filteredChildren = filterChildren(el.props.children);
    newElement = {
      ...el,
      props: {
        ...el.props,
        children: filteredChildren
      }
    };
  }
  return newElement ? cb(newElement) : cb(el);
};

export const handleFilterXss = (el) => {
  if (typeof(el) === "string") return el;
  const props = el.props;
  if(!props) {
    return el;
  }
  const newProps = {
    ...props
  };
  if (newProps.src) {
    newProps.src = xssFilters.uriInDoubleQuotedAttr(props.src);
  }
  return {
    ...el,
    props: newProps
  };
};

const verifyExternalLink = (e, link, confirmWithModal) => {
  e.preventDefault();
  // Does this to prevent xss attacks
  const tmpLink = document.createElement("a");
  tmpLink.href = link;
  const externalLink = (tmpLink.hostname && tmpLink.hostname !== window.top.location.hostname);
  // if this is an external link, show confirmation dialog
  if (externalLink) {
    confirmWithModal(modalTypes.CONFIRM_ACTION, {
      title: "External Link Warning",
      message: (
        <React.Fragment>
          <p style={{ marginBottom: "10px" }}>
            You are about to be sent to an external website! <strong className="red">Do not</strong> enter your Politeia credentials or reveal any other sensitive information.
          </p>
          <p>
            External link: {link}
          </p>
          <p>
            External domain: <strong className="red">{tmpLink.hostname}</strong>
          </p>
          <p style={{ marginTop: "10px" }}>
            Are you sure you want to proceed?
          </p>
        </React.Fragment>
      )
    }).then(confirm => {
      if (confirm) {
        const newWindow = window.open();
        newWindow.opener = null;
        newWindow.location.href = link;
        newWindow.target = "_blank";
      }
    });
  } else if(tmpLink.hostname) {
    window.location.href = link;
  } else {
    console.log("Blocked potentially malicious link: ", link);
  }
};

const imageHandler = (confirmWithModal) => ({ src, alt }) => {
  return <a
    target="_blank"
    rel="nofollow noopener noreferrer"
    onClick={(e) => confirmWithModal && verifyExternalLink(e, src, confirmWithModal)}
    href={src}>{alt}
  </a>;
};

const linkHandler = (confirmWithModal) => ({ href, children }) => {
  return( <a
    target="_blank"
    rel="nofollow noopener noreferrer"
    onClick={(e) => confirmWithModal && verifyExternalLink(e, href, confirmWithModal)}
    href={href}
  >{children[0]}</a>);
};

const rootHandler = (filterXss) => (el) => {
  if(filterXss) {
    el = traverseChildren(el, handleFilterXss);
  }
  const {
    children,
    ...props
  } = el;
  return <div {...props}>{children}</div>;
};

export const customRenderers = (filterXss, confirmWithModal) => {
  const imageRenderer = imageHandler(confirmWithModal);
  const linkRenderer = linkHandler(confirmWithModal);
  const rootRenderer = rootHandler(filterXss);

  return {
    image: imageRenderer,
    imageReference: imageRenderer,
    link: linkRenderer,
    linkReference: linkRenderer,
    root: rootRenderer
  };
};
