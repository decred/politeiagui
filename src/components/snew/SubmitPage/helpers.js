const concatSupportedChars = (supportedChars) =>
  supportedChars.reduce((str, v) => str + v, "");

export const invoiceInstructions = ({
  cmscontactsupportedchars,
  cmsnamelocationsupportedchars,
  maxlocationlength,
  minlocationlength,
  mincontactlength,
  maxcontactlength,
  minnamelength,
  maxnamelength,
  minlineitemcollength,
  maxlineitemcollength,
  invoicefieldsupportedchars
}) => `
***Contractor Name:*** This is whatever name you identify yourself with the DHG, typically something beyond a mere handle or nick. 
   - Allowed chars: ${concatSupportedChars(cmsnamelocationsupportedchars)}
   - Min length: ${minnamelength}
   - Max length: ${maxnamelength}
  
***Contractor Location:*** This is the country you are currently located, or primarily residing.
  - Allowed chars: ${concatSupportedChars(cmsnamelocationsupportedchars)}
  - Min length: ${minlocationlength}
  - Max length: ${maxlocationlength}
  
***Contractor Contact:*** Contact information in case an administrator would need to reach out to discuss something, typically an email address or chat nick.
  - Allowed chars: ${concatSupportedChars(cmscontactsupportedchars)}
  - Min length: ${mincontactlength}
  - Max length: ${maxcontactlength}
  
***Contractor Rate:*** This is the previously agreed upon rate you will be performing work.
  
***Payment Address:*** This is the DCR address where you would like to receive payment.  
  
***Line Items:***
  * Type: Currently can be 1 (Labor), 2 (Expense), 3 (Misc) or 4 (Sub Contractor)
  * Domain: The broad category of work performed/expenses spent (for example, Development, Marketing, Community etc).
  * Subdomain: The specific project or program of which the work or expenses are related (for example, Decrediton, dcrd, NYC Event).
  * Description: A thorough description of the work or expenses.
  * Proposal Token: The proposal token that proposal refers to.
  * Sub Contractor ID: If line item is 4 (Sub Contractor), you must provide who is the Sub Contractor.
  * Sub Contractor Rate: The rate decided upon an agreement between the Supervisor and the Sub Contractor.
  * Labor: The number of hours of work performed.
  * Expenses: The cost of the line item (in USD).
    
Line items policy for domain, subdomain and description:
  - Allowed chars: ${concatSupportedChars(invoicefieldsupportedchars)}
  - Min length: ${minlineitemcollength}
  - Max length: ${maxlineitemcollength}
  `;
