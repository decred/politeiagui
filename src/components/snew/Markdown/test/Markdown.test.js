import React from "react";
import ReactMarkdown from "react-markdown";
import renderer from "react-test-renderer";
import { customRenderers } from "../helpers";

const maliciousBodyText =
  "![clickforxss](javascript:alert('XSS'))  [clickforxss](javascript:alert('XSS'))";

it("filter potencial 'XSS' attackers", () => {
  const tree = renderer
    .create(
      <ReactMarkdown
        source={maliciousBodyText}
        renderers={customRenderers(true)}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
