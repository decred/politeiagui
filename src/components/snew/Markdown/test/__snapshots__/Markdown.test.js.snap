// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter potencial 'XSS' attackers 1`] = `
<div>
  <p>
    <a
      href="x-javascript:alert('XSS')"
      onClick={[Function]}
      rel="nofollow"
    >
      clickforxss
    </a>
      
    <a
      href="javascript:void(0)"
      onClick={[Function]}
      rel="nofollow"
    >
      clickforxss
    </a>
  </p>
</div>
`;
